internal/manager/telemetry/telemetry_test.go

```diff
@@ -0,0 +1,430 @@
+					ObjectMeta: metav1.ObjectMeta{Name: fmt.Sprintf("test-%d", i)},
+					Spec: gatewayv1alpha2.TLSRouteSpec{
+						Rules: []gatewayv1alpha2.TLSRouteRule{{}},
+					},
+				},
+				metav1.CreateOptions{},
+			)
+			require.NoError(t, err)
+
+			_, err = gcl.GatewayV1beta1().ReferenceGrants(namespace).Create(
+				ctx,
+				&gatewayv1beta1.ReferenceGrant{
+					ObjectMeta: metav1.ObjectMeta{Name: fmt.Sprintf("test-%d", i)},
+					Spec: gatewayv1beta1.ReferenceGrantSpec{
+						From: []gatewayv1beta1.ReferenceGrantFrom{
+							{
+								Kind:      "test",
+								Namespace: metav1.NamespaceDefault,
+							},
+						},
+						To: []gatewayv1beta1.ReferenceGrantTo{
+							{
+								Kind: "test",
+							},
+						},
+					},
+				},
+				metav1.CreateOptions{},
+			)
+			require.NoError(t, err)
+		}
+	}
+}
+
+func verifyTelemetryReport(t *testing.T, c *assert.CollectT, k8sVersion *version.Info, report []byte) {
+	t.Helper()
+	hostname, err := os.Hostname()
+	assert.NoError(c, err)
+	semver, err := utilversion.ParseGeneric(k8sVersion.GitVersion)
+	assert.NoError(c, err)
+
+	// Report contains stanza like:
+	// id=57a7a76c-25d0-4394-ab9a-954f7190e39a;
+	// uptime=9;
+	// that is not stable across runs, so we need to remove it.
+	reportToAssert, err := removeStanzaFromReport(string(report), "id")
+	assert.NoError(c, err)
+	reportToAssert, err = removeStanzaFromReport(reportToAssert, "uptime")
+	assert.NoError(c, err)
+
+	assert.Equal(
+		c,
+		fmt.Sprintf(
+			"<14>"+
+				"signal=kic-ping;"+
+				"db=off;"+
+				"feature-combinedroutes=true;"+
+				"feature-combinedservices=false;"+
+				"feature-expressionroutes=false;"+
+				"feature-fillids=false;"+
+				"feature-gateway-service-discovery=false;"+
+				"feature-gateway=false;"+
+				"feature-gatewayalpha=false;"+
+				"feature-knative=false;"+
+				"feature-konnect-sync=false;"+
+				"hn=%s;"+
+				"kv=3.3.0;"+
+				"v=NOT_SET;"+
+				"k8s_arch=%s;"+
+				"k8s_provider=UNKNOWN;"+
+				"k8sv=%s;"+
+				"k8sv_semver=%s;"+
+				"k8s_gatewayclasses_count=2;"+
+				"k8s_gateways_count=4;"+
+				"k8s_grpcroutes_count=4;"+
+				"k8s_httproutes_count=4;"+
+				"k8s_nodes_count=2;"+
+				"k8s_pods_count=4;"+
+				"k8s_referencegrants_count=4;"+
+				"k8s_services_count=5;"+
+				"k8s_tcproutes_count=4;"+
+				"k8s_tlsroutes_count=4;"+
+				"k8s_udproutes_count=4;"+
+				"\n",
+			hostname,
+			k8sVersion.Platform,
+			k8sVersion.GitVersion,
+			"v"+semver.String(),
+		),
+		reportToAssert,
+	)
+}
+
+func removeStanzaFromReport(report string, stanza string) (string, error) {
+	const idStanzaEnd = ";"
+	stanza = stanza + "="
+	start := strings.Index(report, stanza)
+	if start == -1 {
+		return "", errors.New("stanza not found in report")
+	}
+	end := strings.Index(report[start:], idStanzaEnd)
+	if end == -1 {
+		return "", errors.New("stanza end not found in report")
+	}
+	end += start
+	return report[:start] + report[end+1:], nil
+}
+
+func generateSelfSignedCert() (tls.Certificate, error) {
```

Comment:
Is there a chance we reuse what's done in https://github.com/Kong/kubernetes-ingress-controller/blob/22f6bad0f1c25e973aa6def5db79fab7fcf058bb/internal/adminapi/kong_test.go#L121-L147 and use the same helper both here and there?

Reply:
Yeah unification would be good think. I have to take a closer look. I have much simpler use case I've just need a certificate to create TLS server, I don't care about client authentication, etc. What you share is boilerplate to test `mTLS`

Reply:
Can we create an issue to create said test helpers for creating certs? Common parts could be shared there with this and mentioned `kong_test.go`. WDYT?

Reply:
created issue https://github.com/Kong/kubernetes-ingress-controller/issues/4118